README.md
# simple_rich_text

Easily format Flutter text with simple format characters.

Motivation: lowest-possible development friction to add color and formatting to text.

In comparison, [easy_rich_text](https://pub.dev/packages/easy_rich_text) requires lots of code (i.e, patternList of EasyRichTextPattern objects). 



# Format Characters

These are non-standard (*not* markdown compatible) but are more intuitive, in my opinion:

| format character      | format effect                | simple_rich_text string input | Flutter output |
|-----------|----------------------|----------------------|----------------------|
| asterisk (*)       | bold            | "this is \*bold\*" | this is **bold** |
| slash (/)       | italics            | "this is /italicized/" |this is *italicized* |
| underscore (_)       | underline            | "this is \_underlined\_" |this is <ins>underlined</ins> |



## Example Input
```
SimpleRichText(r'*_/this is all three*_/ (*{color:red}bold*, _{color:green}underlined_, and /{color:brown}italicized/). _{push:home;color:blue}clickable hyperlink to home screen_')
```

## Example Flutter Output
![Screenshot](example.png)

 





# Attributes

Attribute pairs are placed in curly brackets immediately after first character marker.
Each pair is separated by a semicolon (;) and can be in any order.
Each pair has syntax *name:value*.



| key      | meaning                | implemented Dart code |
|-----------|----------------------|----------------------|
| color   | red, blue, etc           | textStyle.color: *color-value* |
| pop       | pop the navigation stack            |  Navigator.pop(context);  |
| push       | push the value onto the navigation stack            | Navigator.pushNamed(context, '/$route');  |
| repl       | replace the top route on the navigation stack            | Navigator.popAndPushNamed(context, '/$route');  |


# Colors Supported

Change text color by passing color as attribute:

## Example
```
"this is blue hyperlink to the _{color:blue,push:calendar}calendar_ screen"
```

| color:color_value      | hex value |
|-----------|-----------|
| aqua | 0x00FFFF |
| black | 0x000000 |
| blue | 0x0000FF |
| brown | 0x9A6324 |
| cream | 0xFFFDD0 |
| cyan | 0x46f0f0 |
| green | 0x00FF00 |
| gray | 0x808080 |
| grey | 0x808080 |
| mint | 0xAAFFC3 |
| lavender | 0xE6BEFF |
| new | 0xFFFF00 |
| olive | 0x808000 |
| orange | 0xFFA500 |
| pink | 0xFFE1E6 |
| purple | 0x800080 |
| red | 0xFF0000 |
| silver | 0xC0C0C0 |
| white | 0xFFFFFF |
| yellow |0xFFFF00 |



# Features

- support text hyperlinks to other screens by preceding formatted text with route inside curly brackets:  e.g., "... _{calendar}go to calendar screen_".




# Sample Inputs:
```
'this is /italic/'

'this is *bold*'

'*_/this is all three*_/ (*bold*, _underlined_, and /italicized/)'

'you can quote characters to NOT format a word \*bold\*'

'this is _underline_'

'go to _{/myroute}home_ page'

'this is ~important~(red).'

'this is _*bold and underlined*_.'
```


# Notes

You can use multiple characters at the same time:
```
"this is _/underlined and italicized/_"
```

You can be sloppy!  Unlike HTML, for convenience, if using multiple characters the open and closed sequences, they _don't_ need to be in exact palindrome matching order:

```
"these are */equivalent/* and works without problems."
"these are */equivalent*/ and works without problems."
```


# Requirements
Ancestor MUST have textDirection set (required by internal RichText widget), either through MaterialApp widget or explicitly wrapped by a Directionality widget:
```
Directionality(
    child: SimpleRichText('Peter', term: 't'),
    textDirection: TextDirection.ltr)
```



# Improvements, Bugs, and Deficiencies

- [ ] user-definable format characters
- [ ] user-definable callback functions for custom formatting
- [ ] support non-named navigation (only pushNamed, etc., supported at present)  




# Pull Requests
Pull requests are welcome!



# Examples

See the example directory.

CHANGELOG.md
# Changelog



## Version 2.0.49 (2021-05-22)

- added pre and post parameters to optionally add leading and trailing TextSpan()s to the RichText composite output \[[issue #7](https://github.com/remoteportal/simple_rich_text/issues/7)\]




## Version 2.0.46 (2021-04-22)

- added textScaleFactor parameter \[[issue #6](https://github.com/remoteportal/simple_rich_text/issues/6)\]



## Version 2.0.45 (2021-04-22)

- added textScaleFactor via MediaQuery \[[issue #6](https://github.com/remoteportal/simple_rich_text/issues/6)\]




## Version 2.0.43 (2021-04-21)

- improved README



## Version 2.0.37 (2021-04-21)

- added new line option, some fontStyles, text overflow, max,etc  [pull request #3]
- breaking change: the text parameter is no longer named--it's just the first parameter




## Version 2.0.36 (2021-04-21)

- can now pass optional textAlign parameter to control how text should be aligned horizontally [issue #1]



## Version 2.0.35 (2021-04-20)

- upgraded to Dart 2.0 [issue #4]
- migrated to null-safety
- migrated from RaisedButton to ElevatedButton
- replaced RichText() with Text.rich() [issue #5]



## Version 1.0.33 (2020-02-19)

- improved documentation



## Version 1.0.31 (2020-02-18)

- added color attributes
- added push and pop nagivation attributes



## Version 1.0.28 (2020-02-18)

- fixed route bug



## Version 1.0.24 (2020-02-18)

- fixed route bug



## Version 1.0.22 (2020-02-18)

- fixed lint errors



## Version 1.0.21 (2020-02-18)

- renamed project



## Version 1.0.20 (2020-02-18)

- initial version

test/simple_rich_text_test.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:simple_rich_text/simple_rich_text.dart';

// https://flutter.dev/docs/cookbook/testing/widget/introduction#3-create-a-testwidgets-test

void main() {
  // Define a test. The TestWidgets function also provides a WidgetTester
  // to work with. The WidgetTester allows you to build and interact
  // with widgets in the test environment.
  testWidgets('test the thing!', (WidgetTester tester) async {
    await tester.pumpWidget(Directionality(
        child: SimpleRichText('This is *bold*'),
        textDirection: TextDirection.ltr));
  });
}


test/substring_highlight_test.dart
import 'package:easy_rich_text/easy_rich_text.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter_test/flutter_test.dart';

// https://flutter.dev/docs/cookbook/testing/widget/introduction#3-create-a-testwidgets-test

void main() {
  // Define a test. The TestWidgets function also provides a WidgetTester
  // to work with. The WidgetTester allows you to build and interact
  // with widgets in the test environment.
  testWidgets('test the thing!', (WidgetTester tester) async {
    await tester.pumpWidget(Directionality(
        child: EasyRichText(text: 'This is *bold*'),
        textDirection: TextDirection.ltr));
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:simple_rich_text/simple_rich_text.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(new MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
//    String text = null;
//    String text = 'no format characters';
//    String text = r'format characters: \*\/\_';
//    String text = '_entire string_';
//    String text = '*/_fully loaded!_*/';
//    String text = 'first *bold*.';
//    String text = r'test*\_quoted\_*.';
//    String text = r'don't*close';
//    String text = r'_start at beginning and do not close';
//    String text = r'/Alissa/ is /awesome/!';
//    String text = r'_back__to back_';
//    String text = r'go to _{route}home_ page';
    //String text = '*_/this is all three*_/ (*{color:red}bold*, _underlined_, and /{color:brown}italicized/). _{push:home;color:blue}clickable hyperlink to home screen_';
    //String text = r'This is a text with \nnew line';
    String text =
        r'Now you can change the ~{fontSize:32}size~ of the text. \nInsert a new line.\nChange the ~{backgroundColor:yellow}background color~\nAnd modify more style as: fontFamily, _{decorationColor:blue}decorationColor_, ~{height:3}height~, etc\n\nToo you can open url: _{http:www.google.com;color:blue}go to Google_\nFinaly, you can define textAlign, maxLines and textOverflow';
    return new MaterialApp(
      //set global text scale factor in builder
      builder: (BuildContext context, Widget child) {
        final MediaQueryData data = MediaQuery.of(context);
        return MediaQuery(
          data: data.copyWith(textScaleFactor: 3),
          child: child,
        );
      },

      home: Scaffold(
        // , chars: r"[*]"
//          backgroundColor: Colors.grey[300],
        body: Center(
          child: SimpleRichText(text,
              logIt: false,
              maxLines: 20,
              pre:
                  TextSpan(text: 'PRE', style: TextStyle(color: Colors.purple)),
              post: TextSpan(
                  text: 'POST', style: TextStyle(color: Colors.purple)),
              style: TextStyle(color: Colors.orange),
              textAlign: TextAlign.center,
              textOverflow: TextOverflow.ellipsis,
              textScaleFactor: 1.5),
        ),
      ),
    );
  }
}


lib/easy_rich_text.dart
library easy_rich_text;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:meta/meta.dart';

/// Widget that renders a string with sub-string highlighting.
class EasyRichText extends StatelessWidget {
  EasyRichText({
    @required this.text,
    this.chars,
    this.context,
    this.fussy,
    this.textStyle = const TextStyle(
      color: Colors.black,
    ),
    this.textStyleHighlight = const TextStyle(
      color: Colors.red,
    ),
  });

  final bool fussy;

  final String chars;

  /// The String to be displayed using rich text.
  final String text;

  /// For navigation
  final BuildContext context;

  /// The {TextStyle} of the {EasyRichText.text} that isn't highlighted.
  final TextStyle textStyle;

  /// The {TextStyle} of the {EasyRichText.term}s found.
  final TextStyle textStyleHighlight;

  @override
  Widget build(BuildContext context) {
    if (text == null) {
      return Text('');
    } else if (text.isEmpty) {
      return Text(text);
    } else {
      print('text: $text');
      List<InlineSpan> children = [];

      Set set = Set();

      // split into array
      List<String> spanList = text.split(RegExp(chars ?? r"[*~/_\\]"));
      print("len=${spanList.length}: $spanList");

      if (spanList.length == 1) {
        print("trivial");
        return Text(text);
      } else {
        int i = 0;
        bool acceptNext = true;
        String route;

        void wrap(String v) {
          print("wrap: $v set=$set");
          TextStyle ts = TextStyle(
              decoration: set.contains('_')
                  ? TextDecoration.underline
                  : TextDecoration.none,
              fontStyle:
                  set.contains('/') ? FontStyle.italic : FontStyle.normal,
              fontWeight:
                  set.contains('*') ? FontWeight.bold : FontWeight.normal);
//        TextSpan span = TextSpan(text: v, style: ts);
          if (route != null) {
//          GestureDetector
//        children.add(WidgetSpan(child: Text('****')));
//          children.add(WidgetSpan(
//              child: GestureDetector(
//            child: Text('CLICK'),
//            onTap: () async {
//              print("TAPPED");
//            },
//          )));
            children.add(TextSpan(
                text: v,
//              text: 'TAP',
                // Beware!  This class is only safe because the TapGestureRecognizer is not given a deadline and therefore never allocates any resources.
                // In any other situation -- setting a deadline, using any of the less trivial recognizers, etc -- you would have to manage the gesture recognizer's lifetime
                // and call dispose() when the TextSpan was no longer being rendered.
                // Since TextSpan itself is @immutable, this means that you would have to manage the recognizer from outside
                // the TextSpan, e.g. in the State of a stateful widget that then hands the recognizer to the TextSpan.
                recognizer: TapGestureRecognizer()
                  ..onTap = () {
                    print("TAP");
                    Navigator.pushNamed(context, '/${route}');
                  },
                style: ts));
            route = null;
          } else {
//          children.add(span);
            children.add(TextSpan(text: v, style: ts));
          }
        }

        void toggle(String m) {
          if (m == r'\') {
            String c = text.substring(i + 1, i + 2);
            print("quote: i=$i: $c");
            wrap(c);
            acceptNext = false;
          } else {
            if (acceptNext) {
              if (set.contains(m)) {
                print("REM: $m");
                set.remove(m);
              } else {
                print("ADD: $m");
                set.add(m);
              }
            }

            acceptNext = true;
          }
        }

        for (var v in spanList) {
          print("========== $v ==========");
          if (v.isEmpty) {
            if (i < text.length) {
              String m = text.substring(i, i + 1);
//          print("e: $m");
              toggle(m);
              i++;
            }
          } else {
            int adv = v.length;
            if (v[0] == '{') {
              print("check $v");
              int close = v.indexOf('}');
              if (close > 0) {
                String param = v.substring(1, close);
                print("param2=$param");
                route = param;
                v = v.substring(close + 1);
                print("remaining: $v");
              }
            }
            wrap(v);
            i += adv;
            if (i < text.length) {
              String m = text.substring(i, i + 1);
              print("*** format: $m");
              toggle(m);
              i++;
            }
          }
        }

        if (fussy ?? false && set.isNotEmpty) {
          throw 'easy_rich_text: not closed: $set'; //TODO: throw real error?
        }

        return RichText(text: TextSpan(children: children));
      }
    }
  }
}

//  Widget build(BuildContext context) {
//    if (term.isEmpty) {
//      return Text(text, style: textStyle);
//    } else {
//      String termLC = term.toLowerCase();
//
//      List<InlineSpan> children = [];
//      List<String> spanList = text.toLowerCase().split(termLC);
//      int i = 0;
//      for (var v in spanList) {
//        if (v.isNotEmpty) {
//          children.add(TextSpan(
//              text: text.substring(i, i + v.length), style: textStyle));
//          i += v.length;
//        }
//        if (i < text.length) {
//          children.add(TextSpan(
//              text: text.substring(i, i + term.length),
//              style: textStyleHighlight));
//          i += term.length;
//        }
//      }
//      return RichText(text: TextSpan(children: children));
//    }
//  }
//}


lib/simple_rich_text.dart
library simple_rich_text;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

const Map<String, int> colorMap = {
  'aqua': 0x00FFFF,
  'black': 0x000000,
  'blue': 0x0000FF,
  'brown': 0x9A6324,
  'cream': 0xFFFDD0,
  'cyan': 0x46f0f0,
  'green': 0x00FF00,
  'gray': 0x808080,
  'grey': 0x808080,
  'mint': 0xAAFFC3,
  'lavender': 0xE6BEFF,
  'new': 0xFFFF00,
  'olive': 0x808000,
  'orange': 0xFFA500,
  'pink': 0xFFE1E6,
  'purple': 0x800080,
  'red': 0xFF0000,
  'silver': 0xC0C0C0,
  'white': 0xFFFFFF,
  'yellow': 0xFFFF00
};

Color parseColor(String color) {
//  print("parseColor: $color");
  var v = colorMap[color];
  if (v == null) {
    return Colors.red;
  } else {
//    return Color(v);
//    return Colors.green;
//    int n = Color(v);
    Color out = Color((0xff << 24) | v);
//    print("parseColor: $color => $out");
    return out;
  }
}

/// Widget that renders a string with sub-string highlighting.
class SimpleRichText extends StatelessWidget {
  SimpleRichText(this.text,
      {this.chars,
      this.context,
      this.fussy,
      this.logIt = false,
      this.maxLines,
      this.pre,
      this.post,
      this.style = const TextStyle(),
      this.textAlign,
      this.textOverflow,
      this.textScaleFactor});

  final String? chars;

  /// For navigation
  final BuildContext? context;

  /// Throw exception if tags not closed (e.g., "this is *bold" because no closing * character)
  final bool? fussy;

  /// Pass in true for debugging/logging/trace
  final bool logIt;

  /// An optional maximum number of lines for the text to span, wrapping if necessary.
  /// If the text exceeds the given number of lines, it will be truncated according
  /// to [overflow].
  ///
  /// If this is 1, text will not wrap. Otherwise, text will be wrapped at the
  /// edge of the box.
  final int? maxLines;

  /// optional leading TextSpan
  final TextSpan? pre;

  /// optional trailing TextSpan
  final TextSpan? post;

  /// The {TextStyle} of the {SimpleRichText.text} that isn't highlighted.
  final TextStyle? style;

  /// The String to be displayed using rich text.
  final String text;

  /// How the text should be aligned horizontally.
  final TextAlign? textAlign;

  /// How visual overflow should be handled.
  final TextOverflow? textOverflow;

  /// The number of font pixels for each logical pixel.
  ///
  /// For example, if the text scale factor is 1.5, text will be 50% larger than
  /// the specified font size.
  final double? textScaleFactor;

  @override
  Widget build(BuildContext context) {
    if (text.isEmpty) {
      return Text('');
    } else {
      //print('text: $text');
      List<InlineSpan> children = [];

      if (pre != null) {
        children.add(pre!);
      }

      Set set = Set();

      bool containsNewLine = text.contains(r'\n');
      log('Contains new line: $containsNewLine');
      List<String> linesList = [];
      if (containsNewLine) {
        linesList = text.split(r'\n');
        log("lines=${linesList.length}: $linesList");
      } else {
        linesList.add(text);
      }

      // Apply modifications into all lines
      for (int k = 0; k < linesList.length; k++) {
        log('Line ${k + 1}: ${linesList[k]}');
        // split into array
        List<String> spanList =
            linesList[k].split(RegExp(chars ?? r"[*~/_\\]"));
        log("len=${spanList.length}: $spanList");

        if (spanList.length == 1) {
          log("trivial");
          if (style == null) {
            children.add(TextSpan(text: ''));
            // If no last line:
            if (k < linesList.length - 1) children.add(TextSpan(text: '\n'));
          } else {
            children.add(TextSpan(text: linesList[k], style: style));
            // If no last line:
            if (k < linesList.length - 1) children.add(TextSpan(text: '\n'));
          }
        } else {
          int i = 0;
          bool acceptNext = true;
          String? cmd;

          void wrap(String v) {
            log("wrap: $v set=$set");

            Map<String, String> map = {};

            if (cmd != null) {
              var pairs = cmd.split(';');
              for (var pair in pairs) {
                var a = pair.split(':');
                if (a.length == 2) {
                  map[a[0].trim()] = a[1].trim();
                } else {
                  throw "attribute value is missing a value (e.g., you passed {key} but not {key:value}";
                }
              }
              log("attributes: $map");
            }

            // TextDecorationStyle "values" is ignored
            TextDecorationStyle? _textDecorationStyle;
            if (map.containsKey('decorationStyle')) {
              if (map['decorationStyle'] == 'dashed')
                _textDecorationStyle = TextDecorationStyle.dashed;
              if (map['decorationStyle'] == 'double')
                _textDecorationStyle = TextDecorationStyle.double;
              if (map['decorationStyle'] == 'dotted')
                _textDecorationStyle = TextDecorationStyle.dotted;
              if (map['decorationStyle'] == 'solid')
                _textDecorationStyle = TextDecorationStyle.solid;
              if (map['decorationStyle'] == 'wavy')
                _textDecorationStyle = TextDecorationStyle.wavy;
            }

            TextStyle ts;
            ts = style!.copyWith(
              color: map.containsKey('color')
                  ? parseColor(map['color']!)
                  : style!.color,
              decoration: set.contains('_')
                  ? TextDecoration.underline
                  : TextDecoration.none,
              fontStyle:
                  set.contains('/') ? FontStyle.italic : FontStyle.normal,
              fontWeight:
                  set.contains('*') ? FontWeight.bold : FontWeight.normal,
              fontSize: map.containsKey('fontSize')
                  ? double.parse(map['fontSize']!)
                  : style!.fontSize,
              fontFamily: map.containsKey('fontFamily')
                  ? '${map['fontFamily']}'
                  : style!.fontFamily,
              backgroundColor: map.containsKey('backgroundColor')
                  ? parseColor(map['backgroundColor']!)
                  : style!.backgroundColor,
              decorationColor: map.containsKey('decorationColor')
                  ? parseColor(map['decorationColor']!)
                  : style!.decorationColor,
              decorationStyle: _textDecorationStyle ?? style!.decorationStyle,
              decorationThickness: map.containsKey('decorationThickness')
                  ? double.parse(map['decorationThickness']!)
                  : style!.decorationThickness,
              height: map.containsKey('height')
                  ? double.parse(map['height']!)
                  : style!.height,
              letterSpacing: map.containsKey('letterSpacing')
                  ? double.parse(map['letterSpacing']!)
                  : style!.letterSpacing,
              wordSpacing: map.containsKey('wordSpacing')
                  ? double.parse(map['wordSpacing']!)
                  : style!.wordSpacing,
            );

            if (map.containsKey('pop') ||
                map.containsKey('push') ||
                map.containsKey('repl') ||
                map.containsKey('http')) {
//            print("BBB cmd=$cmd");
//          GestureDetector
//        children.add(WidgetSpan(child: Text('****')));
//          children.add(WidgetSpan(
//              child: GestureDetector(
//            child: Text('CLICK'),
//            onTap: () async {
//              //print("TAPPED");
//            },
//          )));

              // assert(context != null, 'must pass context if using route links');

              onTapNew(String caption, Map m) {
                if (map.containsKey('push')) {
                  String v = map['push']!;
                  return () {
                    log("TAP: PUSH: $caption => /$v");
                    // assert(v != null);
                    Navigator.pushNamed(context, '/$v');
                    // Nav.push('/$v');
                  };
                } else if (map.containsKey('repl')) {
                  String v = map['repl']!;
                  return () {
                    log("TAP: POP&PUSH: $caption => /$v");
                    // assert(v != null);
                    Navigator.popAndPushNamed(context, '/$v');
                    //  Nav.repl('/$v');
                  };
                } else if (map.containsKey('http')) {
                  String v = map['http']!;
                  return () async {
                    log("TAP: HTTP: $caption => /$v");
                    // assert(v != null);
                    try {
                      await launch('http://$v');
                    } catch (e) {
                      log('Could not launch http://$v: $e');
                      try {
                        await launch('https://$v');
                      } catch (e) {
                        log('Could not launch https://$v: $e');
                      }
                    }
                  }; // TODO add possibility of tel, mailto, sms, whats,...?
                } else {
                  return () {
                    log("TAP: $caption => pop");
                    Navigator.pop(context);
                  };
                }
              }

              children.add(TextSpan(
                  text: v,
                  // Beware!  This class is only safe because the TapGestureRecognizer is not given a deadline and therefore never allocates any resources.
                  // In any other situation -- setting a deadline, using any of the less trivial recognizers, etc -- you would have to manage the gesture recognizer's lifetime
                  // and call dispose() when the TextSpan was no longer being rendered.
                  // Since TextSpan itself is @immutable, this means that you would have to manage the recognizer from outside
                  // the TextSpan, e.g. in the State of a stateful widget that then hands the recognizer to the TextSpan.
                  recognizer: TapGestureRecognizer()..onTap = onTapNew(v, map),
                  style: ts));
            } else {
              children.add(TextSpan(text: v, style: ts));
            }
          }

          void toggle(String m) {
            if (m == r'\') {
              String c = linesList[k].substring(i + 1, i + 2);
              log("quote: i=$i: $c");
              wrap(c);
              acceptNext = false;
            } else {
              if (acceptNext) {
                if (set.contains(m)) {
                  log("REM: $m");
                  set.remove(m);
                } else {
                  log("ADD: $m");
                  set.add(m);
                }
              }

              acceptNext = true;
            }
          }

          for (var v in spanList) {
            log("========== $v ==========");
            cmd = null; //TRY
            if (v.isEmpty) {
              if (i < linesList[k].length) {
                String m = linesList[k].substring(i, i + 1);
                toggle(m);
                i++;
              }
            } else {
              int adv = v.length;
              if (v[0] == '{') {
                log("link: $v");
                int close = v.indexOf('}');
                if (close > 0) {
                  cmd = v.substring(1, close);
                  log("AAA cmd=$cmd");
                  v = v.substring(close + 1);
                  log("remaining: $v");
                }
              }
              wrap(v);
              i += adv;
              if (i < linesList[k].length) {
                String m = linesList[k].substring(i, i + 1);
                log("*** format: $m");
                toggle(m);
                i++;
              }
            }
          }

          if ((fussy ?? false) && set.isNotEmpty) {
            throw 'simple_rich_text: not closed: $set'; //TODO: throw real error?
          }

          // If no last line:
          if (k < linesList.length - 1) children.add(TextSpan(text: '\n'));
        }
      }

      if (post != null) {
        children.add(post!);
      }

      return RichText(
          maxLines: this.maxLines ?? null,
          overflow: this.textOverflow ?? TextOverflow.clip,
          text: TextSpan(children: children),
          textAlign: this.textAlign ?? TextAlign.start,
          textScaleFactor:
              this.textScaleFactor ?? MediaQuery.of(context).textScaleFactor);
    } // else
  } // build

  void log(String s) {
    if (logIt) print('---- $s');
  }
} // class
